[Core] Add Retargeting and `SimpleTarget`, consolidate target logic

zbee opened the pull request:
This PR aims to remove all dependency on Redirect/Reaction, clean up and consolidate some of our targeting logic into one place, add some features that previously would have depended on Redirect/Reaction, and obtain parity and customizability between the target that Healing combos check and the target that healing abilities are then cast on.

Original design in [this RFC](https://gist.github.com/zbee/08c26b9b5bb07d57270b999689100033), if you'd like to look at some of my still-pending ideas.

- [X] Implement Action Retargeting
  - [X] `ActionRetargeting` <sup>(~400 of the new lines [and ~400 more of documentation])</sup>
    - [X] `Ability.Retarget(TargetResolver)` saves a retarget order
      - [X] Handles standard Retargeting in a combo
      - [X] Handles Retargeting on a feature that Retargets itself
      - [X] Handles ground-targeted actions
    - [X] Storage of retargets
    - [X] Dropping all retargets on area changes
    - [X] Fast fetching of retargets at action-execution time
    - [X] Limiting to prevent heavy dictionary operation spam
    - [X] Overwriting to allow refreshing of resolvers without having to manually release retargets
    - [X] Age-out old retargets
    - [X] Won't try to use Retargets on abilities that are incompatible with the new target
  - [X] Add Debugging section <sup>(~100 of the new lines)</sup>\
      ![Image of debug section for Retargeting](https://github.com/user-attachments/assets/8b002757-84f8-4ae9-895f-69732af31221)
  - [X] Add documentation section <sup>(~300 of the new lines)</sup>
    - [X] With Example code
  - [X] Move all Reaction/Redirect-dependent options to use Retargeting
- [X] Implement Simple Targets
  - [X] Consolidate as much as makes sense of Targeting stuff into `SimpleTarget` <sup>(~600 of the new and removed lines)</sup>
    - [X] From `Services.PartyUITargeting`\
          (removed, split into `PronounService` and `GameObjectExtensions`, and into `SimpleTarget`, obviously)
    - [X] From `CustomCombo.Functions.Target`
    - [X] Organize `CustomCombo.Functions.Target`
  - [X] Expand to include a lot of [Reaction's Options](https://github.com/UnknownX7/ReAction/blob/913f1d8f1bb07a57f7b7e81e9511ad727d57fc39/PronounManager.cs) <sup>(~300 of the new lines)</sup>\
        (implemented all of them except the ones for each and every job)
  - [X] Move Healing targets to use a stack in `SimpleTarget` <sup>(~300 of the new lines)</sup>
    - [X] Settings for Users to define globally if MouseOver and Focus targets should get included into the default heal stack
    - [X] Remove Healer MouseOver/etc options
    - [X] Remove the corresponding Configs
    - [X] Add optional Retargeting to all healing actions\
          ![Retarget Healing Actions setting](https://github.com/user-attachments/assets/a87b9829-0cae-4e2c-b1d7-6bd0230de74a)
      - [X] Test AST
      - [X] Test WHM
      - [X] Test SGE
      - [X] Test SCH
  - [x] Settings for Users to customize the stack for heal targeting <sup>(~600 of the new lines)</sup>
    - [X] Customizing of the default heal stack\
          ![Default Heal Stack Customization](https://github.com/user-attachments/assets/1ac85802-d115-40ad-8de7-a1d7830d6545)
    - [X] Ability to create a custom heal stack\
        ![Custom Heal Stacks Manager UI](https://github.com/user-attachments/assets/8687027f-43da-4a4c-8054-d8dd08c91594)
    - [X] Custom Heal Stacks resolve to actual `SimpleTarget`s, and replaces the Heal Stack if enabled
- [X] Add new Retargeted features <sup>(~250 of the new lines)</sup>
  - [X] AST Dignity as a stand-alone Feature
  - [X] ALL Healer Rez
  - [X] ALL Mage Rez
  - [X] All Healers' Alternative Rez Features
  - [X] All Casters' Alternative Rez Features
  - [X] WAR Bloodwhetting Sync Feature, which can include `Nascent Flash` if targeting an ally
  - [X] DRK TBN Retargeting Feature
  - [X] DRK Oblation Retargeting Feature
- [X] Convert AST's Card Targeting to be a `TargetResolver`. <sup>(~200 of the new lines, ~400 of the removed lines)</sup>
  - [X] Add option to Quick Target cards that Retargets non-combo cards
- [X] Finish implementation of DNC's Dance Partner features <sup>(~380 of the new and removed lines)</sup>\
      (which require Retargeting and thus were only stubbed out)
- [X] Write a "New Features" popup <sup>(~600 of the new and removed lines)</sup>\
      ![new feature info popup](https://github.com/user-attachments/assets/45007c9e-cf05-4a5b-953b-c2401ce24262)


---

@Taurenkey The changes to your code includes a lot of the new stuff, naturally, but also AST's switch to a simpler and faster `TargetResolver` instead of all the manual targeting and target resolution for cards.
But also it retargets AST ST Healing options to exactly match the Heal Target whose HP was checked for the action.

@Tartarga The changes to your code are to use the new Heal stack Simple Target instead of `GetHealTarget()`, and to move the "Healing Targets should check MouseOver" from being a Config in each Healer's settings to be a global setting. You might want to test it, but it's effectively the same.
Also, it adds Retargeting to SCH and SGE ST Healing options to exactly match the Heal Target whose HP was checked for the action.

@Akechi-kun The changes to your code are just `Target` method renames.\
**There is nothing to test**.

@Kaeris-Tempest The changes to your code is everything `Target`. I removed several dead methods, moved some methods to `SimpleTarget` instead, renamed just a couple methods that had few usages, and reorganized it all into `#regions`.\
**There is nothing to test**.

> - Tested non-Retargeted stuff in Raid in general, so did @Kagekazu, doesn't break everything.
> - DNC Dance Partner Features tested in Fates and Expert spam, and tested by my friend in Raid, doesn't break anything.
> - AST Card Features tested in Expert spam, and by @Cthulhu252 which you can see in the comments below.
> - Heal Retargeting tested in Expert spam, with and without a Custom Heal Stack.
>   - On AST
>   - On WHM
>   - On SGE, Friend on SGE
>   - Friend on SCH

zbee commented:
@Tartarga @Kaeris-Tempest I don't know that y'all have anything pending for it anymore, but I would appreciate if any further Target-focused changes could wait (or be PR'd to my branch).

zbee commented:
@Cthulhu252 This re-implements AST's card targeting, and removes the old stuff. I would really really appreciate if you could pull this and double-check that it picks targets similarly to #521 (which is now the default behavior for the Quick Target Cards Feature).

zbee commented:
@Cthulhu252 would you be able to give any more information on this, and/or confirm that you don't have a Redirect/Reaction configuration that is affecting this?
I cannot replicate the issue in the Single Target or AoE rotations (currently manually playing cards is not retargeted, will be later).

Cthulhu252 commented:
> @Cthulhu252 would you be able to give any more information on this, and/or confirm that you don't have a Redirect/Reaction configuration that is affecting this? I cannot replicate the issue in the Single Target or AoE rotations (currently manually playing cards is not retargeted, will be later).

I was only testing the manually played cards, i don't use the rotations so thats the issue. I see you've pushed sone commits for this so I'll test it again later today

zbee commented:
@Cthulhu252 Ah, okay, yup that'd be it. Those should now be Retargeted, as of dbd2d60.

That setting should default on for you, and is located in the "Quick Target Damage Cards" Feature at the bottom of AST 👍🏻

Cthulhu252 commented:
> That setting should default on for you, and is located in the "Quick Target Damage Cards" Feature at the bottom of AST
@zbee 
Tested again and the manually played cards work and was on by default, and priority logic seems fine. 

Couple of small things, previously the MouseOver Override only worked on the UI element while it now works on mouseover player models, would it be possible to have an option for this to be only UI elements again? It's far too easy to accidently mouseover a player and mess up the targetting while the UI allows for more deliberate choices (e.g. a far more geared player)

Secondly, I see you built a 0.75 second delay in, which is fine for purely auto targetting, but when using either mouseover or hard target override is a bit slow. Could this be dropped to 0.5 or even 0.25?

zbee commented:
Yeah, I was thinking about the field targeting, I'll make that be an extra option.

As for the delay: I didn't think about how that'd affect manual overriding. I don't know that I'll lower that delay, but I think I will move it further down, so it only affects the party list selection, and not the overriding; will try that out and give it back for you to check out.

zbee commented:
@Cthulhu252 I made UI MouseOver separate from "UI or Field" MouseOver, and added a new AST option for that.
I also made the delay essentially not affect the targeting overrides (0.1s), and moved the 0.75second delay down to only affect party searching.

Cthulhu252 commented:
Ast cards are working perfect now! Thanks
